File: test/fixture_loader.py
'''
Fixture loader for jsonsubschema tests
Loads JSON test fixtures for both Python and JavaScript implementations
'''

import json
import os
from pathlib import Path
from typing import Dict, List, Any, Optional


class FixtureLoader:
    """Load JSON test fixtures from the fixtures directory"""

    def __init__(self, fixtures_dir: Optional[str] = None):
        if fixtures_dir is None:
            # Default to test/fixtures relative to this file
            fixtures_dir = Path(__file__).parent / 'fixtures'
        self.fixtures_dir = Path(fixtures_dir)

        if not self.fixtures_dir.exists():
            raise FileNotFoundError(
                f"Fixtures directory not found: {self.fixtures_dir}")

    def load_fixture(self, category: str, filename: str) -> Dict[str, Any]:
        """
        Load a specific fixture file

        Args:
            category: Subdirectory name (api, types, features, integration)
            filename: JSON filename (e.g., 'integer.json')

        Returns:
            Dictionary with 'description' and 'tests' keys
        """
        path = self.fixtures_dir / category / filename
        if not path.exists():
            raise FileNotFoundError(f"Fixture file not found: {path}")

        with open(path, 'r') as f:
            return json.load(f)

    def load_category(self, category: str) -> Dict[str, Dict[str, Any]]:
        """
        Load all fixture files in a category

        Args:
            category: Subdirectory name (api, types, features, integration)

        Returns:
            Dictionary mapping filename to fixture data
        """
        category_path = self.fixtures_dir / category
        if not category_path.exists():
            raise FileNotFoundError(f"Category directory not found: {category_path}")

        fixtures = {}
        for file_path in category_path.glob('*.json'):
            with open(file_path, 'r') as f:
                fixtures[file_path.name] = json.load(f)

        return fixtures

    def load_all_fixtures(self) -> Dict[str, Dict[str, Dict[str, Any]]]:
        """
        Load all fixture files organized by category

        Returns:
            Dictionary with structure: {category: {filename: fixture_data}}
        """
        all_fixtures = {}

        for category_path in self.fixtures_dir.iterdir():
            if category_path.is_dir() and not category_path.name.startswith('.'):
                category = category_path.name
                all_fixtures[category] = self.load_category(category)

        return all_fixtures

    def iter_tests(self, category: Optional[str] = None, filename: Optional[str] = None):
        """
        Iterator that yields individual test cases

        Args:
            category: Optional category to filter by
            filename: Optional filename to filter by

        Yields:
            Tuple of (category, filename, test_case)
        """
        if category and filename:
            # Load specific file
            fixture_data = self.load_fixture(category, filename)
            for test in fixture_data['tests']:
                yield (category, filename, test)
        elif category:
            # Load all files in category
            fixtures = self.load_category(category)
            for fname, fixture_data in fixtures.items():
                for test in fixture_data['tests']:
                    yield (category, fname, test)
        else:
            # Load all fixtures
            all_fixtures = self.load_all_fixtures()
            for cat, fixtures in all_fixtures.items():
                for fname, fixture_data in fixtures.items():
                    for test in fixture_data['tests']:
                        yield (cat, fname, test)


def get_test_cases(category: str, filename: str) -> List[Dict[str, Any]]:
    """
    Convenience function to get test cases from a fixture file

    Args:
        category: Category directory name
        filename: JSON filename

    Returns:
        List of test case dictionaries
    """
    loader = FixtureLoader()
    fixture = loader.load_fixture(category, filename)
    return fixture['tests']

File: js/test/fixture-loader.ts
/**
 * Fixture loader for JavaScript/TypeScript tests
 * Loads JSON test fixtures shared with Python implementation
 */

import { readFileSync, readdirSync, statSync } from 'fs'
import { join, dirname } from 'path'
import { fileURLToPath } from 'url'
import type { FixtureFile, TestCase } from '../src/types'

// Get the directory path for ES modules
const __filename = fileURLToPath(import.meta.url)
const __dirname = dirname(__filename)

export class FixtureLoader {
  private fixturesDir: string

  constructor(fixturesDir?: string) {
    if (fixturesDir) {
      this.fixturesDir = fixturesDir
    } else {
      // Default to test/fixtures relative to the project root
      this.fixturesDir = join(__dirname, '../../test/fixtures')
    }
  }

  /**
   * Load a specific fixture file
   */
  loadFixture(category: string, filename: string): FixtureFile {
    const path = join(this.fixturesDir, category, filename)
    const content = readFileSync(path, 'utf-8')
    return JSON.parse(content) as FixtureFile
  }

  /**
   * Load all fixture files in a category
   */
  loadCategory(category: string): Record<string, FixtureFile> {
    const categoryPath = join(this.fixturesDir, category)
    const files = readdirSync(categoryPath).filter(f => f.endsWith('.json'))

    const fixtures: Record<string, FixtureFile> = {}
    for (const file of files) {
      const content = readFileSync(join(categoryPath, file), 'utf-8')
      fixtures[file] = JSON.parse(content) as FixtureFile
    }

    return fixtures
  }

  /**
   * Load all fixtures organized by category
   */
  loadAllFixtures(): Record<string, Record<string, FixtureFile>> {
    const allFixtures: Record<string, Record<string, FixtureFile>> = {}

    const categories = readdirSync(this.fixturesDir).filter(name => {
      const stat = statSync(join(this.fixturesDir, name))
      return stat.isDirectory() && !name.startsWith('.')
    })

    for (const category of categories) {
      allFixtures[category] = this.loadCategory(category)
    }

    return allFixtures
  }

  /**
   * Iterator that yields individual test cases
   */
  *iterTests(
    category?: string,
    filename?: string
  ): Generator<{ category: string; filename: string; test: TestCase }> {
    if (category && filename) {
      // Load specific file
      const fixtureData = this.loadFixture(category, filename)
      for (const test of fixtureData.tests) {
        yield { category, filename, test }
      }
    } else if (category) {
      // Load all files in category
      const fixtures = this.loadCategory(category)
      for (const [fname, fixtureData] of Object.entries(fixtures)) {
        for (const test of fixtureData.tests) {
          yield { category, filename: fname, test }
        }
      }
    } else {
      // Load all fixtures
      const allFixtures = this.loadAllFixtures()
      for (const [cat, fixtures] of Object.entries(allFixtures)) {
        for (const [fname, fixtureData] of Object.entries(fixtures)) {
          for (const test of fixtureData.tests) {
            yield { category: cat, filename: fname, test }
          }
        }
      }
    }
  }
}

/**
 * Convenience function to get test cases from a fixture file
 */
export function getTestCases(category: string, filename: string): TestCase[] {
  const loader = new FixtureLoader()
  const fixture = loader.loadFixture(category, filename)
  return fixture.tests
}
